index.bs

```diff
@@ -16,13 +16,443 @@ WPT Display: closed
+
+Infra {#infra}
+==============
+
+This specification depends on the Infra standard. [[!INFRA]]
+
+Storage-Access Request Infrastructure {#request-infrastructure}
+===============================================================
+
+In addition to the new headers themselves, this document introduces some new infrastructure to store and convey metadata in the [=user agent=], particularly on a [=request=].
+
+A [=request=] has a boolean [=struct/item=] <dfn for="request">eligible for storage-access</dfn>. This indicates whether the relevant third-party site allows the relevant "<a permission><code>storage-access</code></a>" permission grant to be used to allow access to cookies, if it exists. Unless stated otherwise it is false.
+
+NOTE: The [=user agent=] is expected to read a [=request=]'s [=request/eligible for storage-access=] boolean when attaching cookies to an outgoing [=request=]. It is not yet possible to specify this, since the cookie store has no defined interface that specs can interact with.
+
+NOTE: A [=request=] also has a [=request/client=], which is an [=environment settings object=] (a kind of [=environment=]) and therefore has a [=environment/has storage access=] boolean. That value is distinct from the value of the [=request=]'s [=request/eligible for storage-access=] boolean.
+
+A [=request=] has an associated <dfn for="request">single-hop cache mode</dfn>, whose value is null or a [=request/cache mode=]. 
+
+This document renames a [=request=]'s [=request/cache mode=] field to <dfn for="request">internal cache mode</dfn>.
+
+<div algorithm="request cache mode">
+This document redefines a [=request=]'s <dfn for="request" export>cache mode</dfn> as the cache mode returned by running the following steps, given a [=request=] |request|:
+    1. If |request|'s [=request/single-hop cache mode=] is not null, return |request|'s [=request/single-hop cache mode=].
+    1. Return |request|'s [=request/internal cache mode=].
+</div>
+
+A <dfn>storage access status</dfn> is one of "`none`", "`inactive`", or "`active`".
+
+<div algorithm>
+    The <dfn for=request>storage access status</dfn> of a [=request=] is the [=storage access status=], or null, returned by running the following steps:
+        1. If the user agent's cookie store would attach cookies with the `SameSite=Strict` attribute to |request|, return null. [[!COOKIES]]
+        1. Let |allowed| be a [=boolean=], initially set to the result of determining whether the user agent's cookie store allows unpartitioned cookies to be attached to |request|.
+        1. If |allowed| is true, return "`active`".
+        1. If |request|'s [=request/eligible for storage-access=] is false, then:
+            1. Set |request|'s [=request/eligible for storage-access=] to true.
```

Comment:
This is presented as how to get `storage access status`, but it mutates request, which seems very unexpected to me.  I think this needs either be restructured, or named as more an algorithm rather than described as a getter.  e.g., "to get and update a request's storage access status, do the following"

Reply:
Technically yes, but in principle it doesn't *really* mutate `request`, since the changes are undone by the time this getter returns. I don't think it makes sense to name this "get and update ..." if there are no observable mutations.

Reply:
Still seems weird to me, though I'll defer to domfarolino. Maybe:

Set |allowed| to be the result of determining whether the user agent's cookie store allows unpartitioned cookies to be attached to |request| with [=request/eligible for storage-access=] set to true.

Reply:
I'd be fine with that, if that's admissible in a spec. (WDYT, Dom?)

I haven't seen that kind of `X with Y` syntax used to mean "make a copy of X and set Y to ..." in specs, so I thought the only available options were to mutate the existing `request`, or make a copy of it and mutate that (but I didn't find anywhere else that makes a copy of a `request`).

Reply:
Sure, the browser layer could avoid knowing about `request`s and know about `SiteForCookies` and all the other metadata instead, that's true. Are you saying that I could avoid mutating the `request` if that hypothetical algorithm accepts the metadata itself, rather than the whole `request`?

Reply:
Personally, my preference would be to give the algorithm whatever specific metadata it needs about the request, instead of the request object. That separates things out from the request pretty well, and lets you set this bit separately without actually modifying the request. If that's too cumbersome or if you want to do it in a follow-up, I think it would be fine to do something like @MattMenke2 described above, but I'd word it like:

```
Set |allowed| to the result of determining whether the user agent's cookie store allows unpartitioned cookies to be attached to |request|, if its [=request/eligible for storage-access=] were set to true.
```

Just to make it ultra clear that (a) you're not modifying the request (which isn't a strict requirement, but seems to be peoples' preference here, which is why we're going that route), and (b) is still very implementable. If I were reading this as an implementer and didn't want to modify the request, I'd probably implement the metadata-passing approach.

Reply:
I think what @domfarolino suggests mostly works for with the nit that it's not the `user agent's cookie store` that allows attaching unpartitioned cookies, it is the user agent itself that makes this decision.

I actually don't really care if we pass in the request directly or not, `eligible for storage-access` could just be an additional argument instead of being attached to the request altogether, no?

Reply:
SGTM, I think the metadata approach is clearer as well. Done, ptal.